pkg/kubectl/cmd/log.go

```diff
@@ -164,74 +168,29 @@ func (o *LogsOptions) Validate() error {
 
 // RunLog retrieves a pod log
 func (o *LogsOptions) RunLog() error {
-	pod, err := o.Client.Pods(o.PodNamespace).Get(o.PodName)
+	mapper, typer := o.Factory.Object()
+	infos, err := resource.NewBuilder(mapper, typer, o.Factory.ClientMapperForCommand()).
+		NamespaceParam(o.Namespace).DefaultNamespace().
+		ResourceNames("pods", o.Name).
+		SingleResourceType().
+		Do().Infos()
 	if err != nil {
 		return err
 	}
-
-	// [-c CONTAINER]
-	container := o.ContainerName
-	if len(container) == 0 {
-		// [CONTAINER] (container as arg not flag) is supported as legacy behavior. See PR #10519 for more details.
-		if len(pod.Spec.Containers) != 1 {
-			podContainersNames := []string{}
-			for _, container := range pod.Spec.Containers {
-				podContainersNames = append(podContainersNames, container.Name)
-			}
-
-			return fmt.Errorf("Pod %s has the following containers: %s; please specify the container to print logs for with -c", pod.ObjectMeta.Name, strings.Join(podContainersNames, ", "))
-		}
-		container = pod.Spec.Containers[0].Name
-	}
-
-	logOptions := &api.PodLogOptions{
-		Container:  container,
-		Follow:     o.Follow,
-		Previous:   o.Previous,
-		Timestamps: o.Timestamps,
-	}
-	if o.SinceSeconds > 0 {
-		// round up to the nearest second
-		sec := int64(math.Ceil(float64(o.SinceSeconds) / float64(time.Second)))
-		logOptions.SinceSeconds = &sec
-	}
-	logOptions.SinceTime = o.SinceTime
-	if o.LimitBytes != 0 {
-		i := int64(o.LimitBytes)
-		logOptions.LimitBytes = &i
-	}
-	if o.Tail >= 0 {
-		i := int64(o.Tail)
-		logOptions.TailLines = &i
+	if len(infos) != 1 {
+		return errors.New("expected a resource")
 	}
+	info := infos[0]
 
-	return handleLog(o.Client, o.PodNamespace, o.PodName, logOptions, o.Out)
+	return handleLog(o.Factory, info.Object, o.Options, o.Out)
 }
 
-func handleLog(client *client.Client, namespace, podID string, logOptions *api.PodLogOptions, out io.Writer) error {
-	// TODO: transform this into a PodLogOptions call
-	req := client.RESTClient.Get().
-		Namespace(namespace).
-		Name(podID).
-		Resource("pods").
-		SubResource("log").
-		Param("follow", strconv.FormatBool(logOptions.Follow)).
-		Param("container", logOptions.Container).
-		Param("previous", strconv.FormatBool(logOptions.Previous)).
-		Param("timestamps", strconv.FormatBool(logOptions.Timestamps))
-
-	if logOptions.SinceSeconds != nil {
-		req.Param("sinceSeconds", strconv.FormatInt(*logOptions.SinceSeconds, 10))
-	}
-	if logOptions.SinceTime != nil {
-		req.Param("sinceTime", logOptions.SinceTime.Format(time.RFC3339))
-	}
-	if logOptions.LimitBytes != nil {
-		req.Param("limitBytes", strconv.FormatInt(*logOptions.LimitBytes, 10))
-	}
-	if logOptions.TailLines != nil {
-		req.Param("tailLines", strconv.FormatInt(*logOptions.TailLines, 10))
+func handleLog(f *cmdutil.Factory, object, options runtime.Object, out io.Writer) error {
```

Comment:
It's reused in a couple of places in `kubectl run` so it's better to leave it as is.